client/actions/projects.js

```diff
@@ -106,20 +119,47 @@ export function updateAndSave(data) {
   };
 }
 
-const doConditionsUpdate = (data, dispatch, state) => {
-  const { application: { basename } } = state;
+const shouldSync = (state, fields = []) => {
+  if (state.application.isSyncing) {
+    return false;
+  }
+  const hasDiff = !isEqual(pick(state.savedProject, fields), pick(state.project, fields));
+  return hasDiff;
+};
+
+const sync = (dispatch, getState, fields, extra = {}) => {
+  const state = getState();
+
+  if (!shouldSync(state, fields)) {
+    return Promise.resolve();
+  }
+
+  dispatch(isSyncing());
+
   const params = {
     method: 'PUT',
-    url: `${basename.replace(/\/edit?/, '')}/conditions`,
-    data
-  };
+    url: `${state.application.basename.replace(/\/edit?/, '')}/conditions`,
+    data: {
+      ...pick(state.project, fields),
+      ...extra
+    }
+  }
+
   return Promise.resolve()
     .then(() => sendMessage(params))
+    .then(() => dispatch(doneSyncing()))
+    .then(() => dispatch(updateSavedProject(state.project)))
+    .then(() => sync(dispatch, getState, fields, extra))
     .catch(err => {
       console.error(err);
-      dispatch(throwError('Error updating conditions'));
+      dispatch(doneSyncing());
+      return sync(dispatch, getState, fields, extra);
     });
-}
+};
+
+const debouncedSync = debounce((dispatch, getState, extra) => {
+  return sync(dispatch, getState, ['conditions', 'retrospectiveAssessment'], extra)
+}, 1000, { maxWait: 5000 });
```

Comment:
Does this mean we do a max of 1 sync per second, with a 5 second timeout?

Reply:
it means it waits for a 1 second pause before trying to sync again, up to a max of 5 seconds. Once the sync is called it starts the recursive dance